### PR TITLE
Gh-1346 - Function Validation

### PR DESCRIPTION
**Pull request by m607123**
Validation added for the Aggregate, Filter, and Transform functions.
Now requires:
An operation that is not null
A schema that is not null
For each element group defined in the "Edges" or "Entities" map of the operation, they must be defined in the schema
Functions to execute should be not null
Any operations/predicates should be applicable to the property upon which they are to be executed

### CODE REVIEW

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Function.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package uk.gov.gchq.gaffer.operation.impl.function;
+
+import uk.gov.gchq.gaffer.data.element.Element;
+import uk.gov.gchq.gaffer.operation.Operation;
+import uk.gov.gchq.gaffer.operation.io.InputOutput;
+import uk.gov.gchq.gaffer.operation.io.MultiInput;
+
+import java.util.Map;
+
+public abstract class Function implements
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Function.java]: Please could you make this an interface we are trying to avoid lots of abstract operation classes.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementFilter.java [hunk 2]
```diff
@@ -31,7 +31,7 @@
 import java.util.function.Predicate;
 
 /**
- * An {@code ElementFiler} is a {@link Predicate} which evaluates a condition against
+ * An <code>ElementFilter</code> is a {@link Predicate} which evaluates a condition against
```
Comment (p013570) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ElementFilter.java]: Please can you use {@code }

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FilterHandler.java [hunk 3]
```diff
@@ -50,6 +51,13 @@
             operation.setEdges(edgeMap);
         }
 
+        final FunctionValidator<Filter> validator = new FunctionValidator<>();
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FilterHandler.java]: Rather than creating a new instance of this for every query, the validator could be a 'private final' field.

**Comment by m607123** [comment]
Closing while changes made

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Function.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.operation.impl.function;
+
+import java.util.Map;
+
+public interface Function {
```
Comment (t616178) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Function.java]: Missing class level and method level javadoc.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Filter.java [hunk 5]
```diff
@@ -39,10 +39,10 @@
  * For more complex queries, a {@link Map} of groups to {@link ElementFilter}s can also be provided to either
  * {@link  uk.gov.gchq.gaffer.data.element.Edge}s or {@link uk.gov.gchq.gaffer.data.element.Entity}s.
  */
-public class Filter implements
+public class Filter implements Function,
         Operation,
         InputOutput<Iterable<? extends Element>, Iterable<? extends Element>>,
-        MultiInput<Element> {
+        MultiInput<Element>  {
```
Comment (t616178) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Filter.java]: Try to avoid adding extra unnecessary whitespace.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Filter.java [hunk 6]
```diff
@@ -39,10 +39,10 @@
  * For more complex queries, a {@link Map} of groups to {@link ElementFilter}s can also be provided to either
  * {@link  uk.gov.gchq.gaffer.data.element.Edge}s or {@link uk.gov.gchq.gaffer.data.element.Entity}s.
  */
-public class Filter implements
+public class Filter implements Function,
         Operation,
```
Comment (t616178) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/function/Filter.java]: This is a redundant interface implementation, since the InputOutput<?> interface extends Operation.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+        if (null != aggregator) {
+            final List<TupleAdaptedBinaryOperator<String, ?>> components = aggregator.getComponents();
+
+            for (final TupleAdaptedBinaryOperator<String, ?> component : components) {
+                final String[] selection = component.getSelection();
+
+                for (final SchemaElementDefinition elementDef : elements.values()) {
+                    final Class[] selectionClasses = Arrays.stream(selection).map(elementDef::getPropertyClass).toArray(Class[]::new);
+                    final Map<String, String> properties = elementDef.getPropertyMap();
+                    if (properties.size() > 0) {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java]: We should prefer the use of ```!Map.isEmpty()``` here.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+                }
+            }
+        } else {
+            for (final SchemaElementDefinition elementDef : elements.values()) {
+                final List<TupleAdaptedBinaryOperator<String, ?>> components = elementDef.getOriginalAggregateFunctions();
+                for (final TupleAdaptedBinaryOperator<String, ?> component : components) {
+                    final String[] selection = component.getSelection();
+                    final Class[] selectionClasses = Arrays.stream(selection).map(elementDef::getPropertyClass).toArray(Class[]::new);
+                    final Map<String, String> properties = elementDef.getPropertyMap();
+                    if (properties.size() > 0) {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java]: We should prefer the use of ```!Map.isEmpty()``` here.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+        final ValidationResult result = new ValidationResult();
+
+        final List<TupleAdaptedPredicate<String, ?>> components = filter.getComponents();
+        for (final TupleAdaptedPredicate<String, ?> component : components) {
+            final String[] selection = component.getSelection();
+
+            for (final SchemaElementDefinition elementDef : elements.values()) {
+                final Class[] selectionClasses = Arrays.stream(selection).map(elementDef::getPropertyClass).toArray(Class[]::new);
+                final Map<String, String> properties = elementDef.getPropertyMap();
+                if (properties.size() > 0) {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java]: We should prefer the use of ```!Map.isEmpty()``` here.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+        final List<TupleAdaptedFunction<String, ?, ?>> components = transformer.getComponents();
+        for (final TupleAdaptedFunction<String, ?, ?> component : components) {
+            final String[] selection = component.getSelection();
+            final String[] projection = component.getProjection();
+
+            for (final SchemaElementDefinition elementDef : elements.values()) {
+                final Class[] selectionClasses = Arrays.stream(selection).map(elementDef::getPropertyClass).toArray(Class[]::new);
+                final Class[] projectionClasses = Arrays.stream(projection).map(elementDef::getPropertyClass).toArray(Class[]::new);
+                final Map<String, String> properties = elementDef.getPropertyMap();
+                if (properties.size() > 0) {
```
Comment (t616178) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java]: We should prefer the use of ```!Map.isEmpty()``` here.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BinaryOperator;
+import java.util.stream.Collectors;
+
+public class FunctionValidator<T extends Function> {
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java]: A lot of the logic in this class is duplicated in ViewValidator. Could you make a utility class for the common logic?

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+        }
+
+        result.add(validateOperation(operation, schema));
+
+        return result;
+    }
+
+    private ValidationResult validateOperation(final T operation, final Schema schema) {
+        final ValidationResult result = new ValidationResult();
+        final boolean opIsAggregate = operation instanceof Aggregate;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/function/FunctionValidator.java]: Rather than doing this, it would be better to have separate classes: AggregateValidator, FilterValidator and TransformFilter, which extends FunctionValidator.

**Comment by m607123** [comment]
Closing while changes made

**Comment by p013570** [comment]
When using the Aggregator:

If you don't supply any groups then it should default to using all groups, similar to the Filter Operation (perhaps also check the Transform Operation)
If you aggregate a transient property then the validation fails as it can't find the class of the transient property. Perhaps just skip the validation for transient properties.
I don't think we need the 'else' block on line 138 of the AggregateValidator.

Please can you check all 3 operations in the REST API too.

**Comment by p013570** [comment]
Closing while changes are made.

**Comment by m607123** [comment]
Checked working as intended in the REST API, with the Road Traffic Demo.
